arduino/discovery/discoverymanager/discoverymanager.go

```diff
@@ -60,233 +69,139 @@ func (dm *DiscoveryManager) IDs() []string {
+		dm.watchersMutex.Unlock()
+		close(watcher.feed)
+	}
+	dm.watchersMutex.Lock()
+	dm.watchers[watcher] = true
+	dm.watchersMutex.Unlock()
+	return watcher, nil
 }
 
-// QuitAll quits all the discoveries managed by this DiscoveryManager.
-// Returns an error for each discovery that fails quitting
-func (dm *DiscoveryManager) QuitAll() []error {
-	errs := dm.parallelize(func(d *discovery.PluggableDiscovery) error {
-		if d.State() == discovery.Dead {
-			// Stop! Stop! It's already dead!
-			return nil
+func (dm *DiscoveryManager) startDiscovery(d *discovery.PluggableDiscovery) (discErr error) {
+	defer func() {
+		if discErr != nil {
+			logrus.Errorf("Discovery %s failed to run: %s", d.GetID(), discErr)
 		}
+	}()
 
-		d.Quit()
-		return nil
-	})
-	return errs
-}
-
-// List returns a list of available ports detected from all discoveries
-// and a list of errors for those discoveries that returned one.
-func (dm *DiscoveryManager) List() ([]*discovery.Port, []error) {
-	var wg sync.WaitGroup
-	// Use this struct to avoid the need of two separate
-	// channels for ports and errors.
-	type listMsg struct {
-		Err  error
-		Port *discovery.Port
+	if err := d.Run(); err != nil {
+		return fmt.Errorf(tr("discovery %[1]s process not started: %[2]w"), d.GetID(), err)
 	}
-	msgChan := make(chan listMsg)
-	dm.discoveriesMutex.Lock()
-	discoveries := []*discovery.PluggableDiscovery{}
-	for _, d := range dm.discoveries {
-		discoveries = append(discoveries, d)
-	}
-	dm.discoveriesMutex.Unlock()
-	for _, d := range discoveries {
-		wg.Add(1)
-		go func(d *discovery.PluggableDiscovery) {
-			defer wg.Done()
-			if d.State() != discovery.Running {
-				// Discovery is not running, it won't return anything
-				return
-			}
-			ports, err := d.List()
-			if err != nil {
-				msgChan <- listMsg{Err: fmt.Errorf(tr("listing ports from discovery %[1]s: %[2]w"), d.GetID(), err)}
-			}
-			for _, p := range ports {
-				msgChan <- listMsg{Port: p}
-			}
-		}(d)
+	eventCh, err := d.StartSync(5)
+	if err != nil {
+		return fmt.Errorf(tr("start syncing discovery %[1]s: %[2]w"), d.GetID(), err)
 	}
 
+	// XXX do better cleanup if the discovery fails to start
+
 	go func() {
-		// Close the channel only after all goroutines are done
-		wg.Wait()
-		close(msgChan)
+		for ev := range eventCh {
+			dm.feed <- ev
+		}
 	}()
+	return nil
+}
 
-	ports := []*discovery.Port{}
-	errs := []error{}
-	for msg := range msgChan {
-		if msg.Err != nil {
-			errs = append(errs, msg.Err)
-		} else {
-			ports = append(ports, msg.Port)
+func (dm *DiscoveryManager) feeder() {
+	// Feed all watchers with data coming from the discoveries
+	for ev := range dm.feed {
+		dm.watchersMutex.Lock()
+		for watcher := range dm.watchers {
+			select {
+			case watcher.feed <- ev:
+				// OK
+			default:
+				// If the watcher is not able to process event fast enough
+				// remove the watcher from the list of watchers
+				go watcher.Close()
+			}
 		}
+		dm.cacheEvent(ev)
+		dm.watchersMutex.Unlock()
```

Comment:
I would use a `defer`

Reply:
In this case, it is not possible because `defer` runs at the function `return` and we are in a `for` loop.

Reply:
Ups you are right, I missed the for loop